language/vm/vm-runtime/src/gas_meter.rs

```diff
@@ -2,19 +2,89 @@
 // SPDX-License-Identifier: Apache-2.0
 
 //! Gas metering logic for the Move VM.
-use crate::{interpreter::InterpreterForGasCost, loaded_data::function::FunctionReference};
+use crate::{
+    code_cache::module_cache::ModuleCache,
+    data_cache::RemoteCache,
+    identifier::{create_access_path, resource_storage_key},
+    interpreter::InterpreterForGasCost,
+    loaded_data::function::FunctionReference,
+};
 use libra_types::{
     account_address::ADDRESS_LENGTH,
+    account_config,
+    identifier::Identifier,
+    language_storage::ModuleId,
     transaction::MAX_TRANSACTION_SIZE_IN_BYTES,
-    vm_error::{StatusCode, VMStatus},
+    vm_error::{sub_status, StatusCode, VMStatus},
+};
+use vm::{
+    access::ModuleAccess,
+    errors::VMResult,
+    file_format::{Bytecode, StructDefinitionIndex},
+    gas_schedule::*,
 };
-use vm::{access::ModuleAccess, errors::VMResult, file_format::Bytecode, gas_schedule::*};
+
+//***************************************************************************
+// Gas Schedule Loading
+//***************************************************************************
+
+lazy_static! {
+    /// The ModuleId for the gas schedule module
+    pub static ref GAS_SCHEDULE_MODULE: ModuleId =
+        { ModuleId::new(account_config::core_code_address(), Identifier::new("GasSchedule").unwrap()) };
+}
+
+pub(crate) fn load_gas_schedule<'alloc, P>(
+    module_cache: P,
+    data_view: &dyn RemoteCache,
+) -> VMResult<CostTable>
+where
+    P: ModuleCache<'alloc>,
+{
+    let address = account_config::association_address();
+    let gas_module = module_cache
+        .get_loaded_module(&GAS_SCHEDULE_MODULE)?
+        .ok_or_else(|| {
+            VMStatus::new(StatusCode::GAS_SCHEDULE_ERROR)
+                .with_sub_status(sub_status::GSE_UNABLE_TO_LOAD_MODULE)
+        })?;
+    // NB: The gas schedule resource is at the 1'st struct handle index in the module -- in
+    // particular since it is defined after the GasCost struct.
+    let struct_tag = resource_storage_key(
+        gas_module,
+        StructDefinitionIndex::new(GAS_SCHEDULE_RESOURCE_DEF_IDX),
+    );
+    let access_path = create_access_path(&address, struct_tag);
+
+    let data_blob = data_view
+        .get(&access_path)
+        .map_err(|_| {
+            VMStatus::new(StatusCode::GAS_SCHEDULE_ERROR)
+                .with_sub_status(sub_status::GSE_UNABLE_TO_LOAD_RESOURCE)
+        })?
+        .ok_or_else(|| {
+            VMStatus::new(StatusCode::GAS_SCHEDULE_ERROR)
+                .with_sub_status(sub_status::GSE_UNABLE_TO_LOAD_RESOURCE)
+        })?;
+    let table: CostTable = lcs::from_bytes(&data_blob).map_err(|_| {
```

Comment:
this another pattern that I genuinely dislike and I am just going to put a comment there. I understand we do it in many different places so it may be a lost battle (or war).
We are taking a Move representation and mapping it straight to a Rust struct and that to me that is a recipe for disaster. Maybe this is more control than other places, maybe all places are well controlled and I don't know about it, but it seems to me we should go to the Move struct and map it with code to the Rust one. Alternatively a solid story which could hopefully give compiler errors is a good way to go.
If I understand it correctly (please correct me) this will happily crash or lead to type holes if things are changes without taking care of the mapping. 
Is my understanding correct?

Reply:
I agree with this.. I don't like it either (but, was replicating what we were doing elsewhere).

I'll take a look at doing the Move -> VM Move value representation -> Rust representation transformation and see how that goes.